### PR TITLE
Use u8 pixels for sequences of bitdepth 8

### PR DESCRIPTION
**Pull request by barrbrain**
Fix all of the paths that do not currently support u8 and at least match the performance of u16.

### CODE REVIEW

File: src/me.rs [hunk 1]
```diff
@@ -97,18 +127,58 @@ mod nasm {
     sum
   }
 
+  #[target_feature(enable = "sse2")]
+  unsafe fn sad_sse2<T: Pixel>(
+    plane_org: &PlaneSlice<'_, T>, plane_ref: &PlaneSlice<'_, T>, blk_h: usize,
+    blk_w: usize
+  ) -> u32 {
+    assert!(mem::size_of::<T>() == 1, "only implemented for u8 for now");
```
Comment (tdaede) [inline src/me.rs]: Uhhhh this is pretty weird. Is there syntax for writing this for just one type? E.g. sad_sse2<T: u8>? @lu-zero ?

Comment (lu-zero) [reply src/me.rs]: I can try to look it up a shorter way but moving it to a trait seems the simplest way to achieve what you want.

Comment (rom1v) [reply src/me.rs]: > Is there syntax for writing this for just one type? E.g. sad_sse2<T: u8>?

If the function supports only `u8` temporarily (i.e. it will also support `u16` in the future), this assert seems good.

On the contrary, if it will never support `u16`, it needs not be generic over `T`:

```rust
  unsafe fn sad_sse2_u8(
    plane_org: &PlaneSlice<'_, u8>, plane_ref: &PlaneSlice<'_, u8>, blk_h: usize,
    blk_w: usize
  ) -> u32 {
    // …
  }
```

However, it is not possible to select the implementation based on the parameter type, like in C++:

```cpp
#include <iostream>

template <typename T>
void print(T value);

template <>
void print(int value) {
    std::cout << "int: " << value << std::endl;
}

template <>
void print(double value) {
    std::cout << "double: " << value << std::endl;
}

int main() {
    print(42);
    print(42.);
}
```

```
$ ./test
int: 42
double: 42
```

The selection of the correct implementation based on the generic type can be done via an explicit test (e.g. `mem::size_of::<T>()`, or via additional methods on the `Pixel` trait).

File: src/cdef.rs [hunk 2]
```diff
@@ -128,10 +127,9 @@ fn constrain(diff: i32, threshold: i32, damping: i32) -> i32 {
 // of the 2-pixel padding around the block, not the block itself.
 // The destination is unpadded.
 unsafe fn cdef_filter_block<T: Pixel>(
-  dst: *mut T, dstride: isize, input: *const T, istride: isize, pri_strength: i32,
+  dst: *mut T, dstride: isize, input: *const u16, istride: isize, pri_strength: i32,
```
Comment (ycho) [inline src/cdef.rs]: Are we then roll back to Monty's original code that always used 16bit for input? I remember you have rescued him when all those Pixel Type changes (by rom1v) has perplexed him with haunting rebasing the other day.

File: src/rdo.rs [hunk 3]
```diff
@@ -1271,19 +1271,18 @@ pub fn rdo_loop_decision<T: Pixel>(sbo: &SuperBlockOffset, fi: &FrameInvariants<
...
-    lrf_output = cdef_sb_frame(fi, &fs.rec);
-  }
+  assert!(fi.sequence.enable_cdef);
+  // all stages; reconstruction goes to cdef so it must be additionally padded
+  // TODO: use the new plane padding mechanism rather than this old kludge.  Will require
+  // altering CDEF code a little.
+  cdef_input = cdef_sb_padded_frame_copy(fi, sbo, &fs.rec, 2);
+  lrf_input = cdef_sb_frame(fi, &fs.rec);
+  lrf_output = cdef_sb_frame(fi, &fs.rec);
+  // FIXME: T:T padded frame copy required for CDEF disabled path.
```
Comment (ycho) [inline src/rdo.rs]: So, CDEF disabled path will be handled in separate patch?
Then, CDEF cannot be disabled with this patch, correct?

File: src/cdef.rs (hunk 2 above)
Comment (barrbrain) [reply src/cdef.rs]: The input code should be much as it was before the Pixel type changes. On the output side, we now additionally have 8bit via the Pixel type.

File: src/rdo.rs (hunk 3 above)
Comment (barrbrain) [reply src/rdo.rs]: Yes, CDEF cannot be disabled after this patch. As the code was, the CDEF direction search is run even when CDEF is disabled. I'd like to follow up in a patch that enables toggling CDEF at runtime.

Comment (tdaede) [reply src/rdo.rs]: Yeah,  the cdef disabling wasn't wired up to any speed level anyway.

File: src/plane.rs [hunk 4]
```diff
@@ -52,21 +53,41 @@ impl<T: Pixel> Debug for Plane<T>
...
+  #[cfg(windows)]
   const STRIDE_ALIGNMENT_LOG2: usize = 4;
+  #[cfg(not(windows))]
+  const STRIDE_ALIGNMENT_LOG2: usize = 5;
 
   /// Data alignment in bytes.
+  #[cfg(windows)]
   const DATA_ALIGNMENT_LOG2: usize = 4;
+  #[cfg(not(windows))]
+  const DATA_ALIGNMENT_LOG2: usize = 5;
```
Comment (tdaede) [inline src/plane.rs]: Is this because we aren't executing AVX2 on Windows? Not sure I'd bother with this considering we'd want to eventually have it on Windows anyway.

Comment (barrbrain) [reply src/plane.rs]: This is because even without any assembly, the win64 binary segfaults with 32-byte alignment. Deferring analysis of that for when we enable nasm there.
